pkg/auth/oauth2_client.go

```diff
@@ -0,0 +1,191 @@
+	// client to the auth server as required for PKCE.
+	PkceCodeChallenge string
+
+
+	TokenChannel chan string
+)
+
+// The following sets up the requirements for generating a standards compliant PKCE code verifier.
+const codeVerifierLenMin = 43
+const codeVerifierLenMax = 128
+const codeVerifierAllowedLetters = "abcdefghijklmnopqrstuvwxyz0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ._~"
+
+// generateCodeVerifier provides an easy way to generate an n-length randomised
+// code verifier.
+func generateCodeVerifier(n int) string {
+	// Enforce standards compliance...
+	if n < codeVerifierLenMin {
+		n = codeVerifierLenMin
+	}
+	if n > codeVerifierLenMax {
+		n = codeVerifierLenMax
+	}
+
+	// Randomly choose some allowed characters...
+	b := make([]byte, n)
+	for i := range b {
+		// ensure we use non-deterministic random ints.
+		j, _ := rand.Int(rand.Reader, big.NewInt(int64(len(codeVerifierAllowedLetters))))
+		b[i] = codeVerifierAllowedLetters[j.Int64()]
+	}
+
+	return string(b)
+}
+
+// generateCodeChallenge returns a standards compliant PKCE S(HA)256 code
+// challenge.
+func generateCodeChallenge(codeVerifier string) string {
+	// Create a sha-265 hash from the code verifier...
+	s256 := sha256.New()
+	s256.Write([]byte(codeVerifier))
+
+	// Then base64 encode the hash sum to create a code challenge...
+	return base64.RawURLEncoding.EncodeToString(s256.Sum(nil))
+}
+
+// resetPKCE cleans up PKCE details and returns the code verifier.
+func resetPKCE() (codeVerifier string) {
+	codeVerifier = PkceCodeVerifier
+	PkceCodeVerifier = ""
+	return codeVerifier
+}
+
+
+// A valid oauth2 client (check the store) that additionally requests an OpenID Connect id token
+var clientConf = goauth.Config{
+	ClientID:     "flytectl",
+	ClientSecret: "foobar",
+	RedirectURL:  "http://localhost:3846/callback",
+	Scopes:       []string{"all"},
+	Endpoint: goauth.Endpoint{
+		TokenURL: "http://localhost:8088/oauth2/token",
+		AuthURL:  "http://localhost:8088/oauth2/authorize",
+	},
+}
+
+type FlyteCtlTokenSource struct {
+	accessToken string
+}
+
+func (ts *FlyteCtlTokenSource) Token() (*goauth.Token, error) {
+	t := &goauth.Token{
+		AccessToken: ts.accessToken,
+		Expiry:      time.Now().Add(1 * time.Minute),
+		TokenType:   "bearer",
+	}
+	return t, nil
+}
+
+type InsecurePerRPCCredentials struct{
+	goauth.TokenSource
+}
+
+func (cr InsecurePerRPCCredentials) GetRequestMetadata(ctx context.Context, uri ...string) (map[string]string, error) {
+	token, err := cr.Token()
+	if err != nil {
+		return nil, err
+	}
+	return map[string]string{
+		"authorization": token.Type() + " " + token.AccessToken,
+	}, nil
+}
+
+func (cr InsecurePerRPCCredentials) RequireTransportSecurity() bool {
+	return false
+}
+
+func StartAuthFlow(ctx context.Context) (grpc.CallOption, error) {
+	// ### oauth2 client ###
+	http.HandleFunc("/callback", CallbackHandler(clientConf)) // the oauth2 callback endpoint
+	port := "3846"
```

Comment:
There is another endpoint in admin `<adminUrl>/config/v1/flyte_client` that gives you the client_id, redirect_uri, scopes to use/request: https://github.com/flyteorg/flyteadmin/blob/oauth2/pkg/rpc/config/flyte_client.go#L23-L28
This way, the administrator can dictate what the client should use... this is required in case of Okta where we can not hardcode our own clientIds and each administrator will need to generate new ones on their own...

Reply:
Done